torch/csrc/profiler/collection.cpp

```diff
@@ -822,147 +822,225 @@ trace_ptr_t addKinetoEvents(
-            // Handle deallocation
-            if (alloc_op.alloc_size_ < 0) {
-              live_storage.erase(StorageImplData(alloc_op.ptr_));
+  for (auto& result : sorted_results) {
+    result->visit(c10::overloaded(
+        [&](ExtraFields<EventType::TorchOp>& torch_op) {
+          for (auto& t : torch_op.inputs_.tensor_metadata_) {
+            if (t.has_value()) {
+              insert_tensor(*t);
             }
-          },
-          [&](ExtraFields<EventType::PyCall>& py_call) {
-            // torch.nn.Module
-            if (py_call.module_.has_value()) {
-              for (auto& p : py_call.module_->parameters_) {
-                insert_tensor(p.metadata_);
-                if (p.grad_metadata_.has_value()) {
-                  insert_tensor(*p.grad_metadata_);
-                }
+          }
+        },
+        [&](ExtraFields<EventType::Allocation>& alloc) {
+          // We won't know which allocations are for Tensor storage yet.
+          // We'll filter after we see all of the op inputs.
+          raw_tensors.push_back(
+              {{NoTensorImpl, 0},
+               {StorageImplData(alloc.ptr_), 0},
+               alloc.id_,
+               alloc.alloc_size_ > 0 ? PairType::kAllocation
+                                     : PairType::kFree});
+        },
+        [&](ExtraFields<EventType::PyCall>& py_call) {
+          // torch.nn.Module
+          if (py_call.module_.has_value()) {
+            for (auto& p : py_call.module_->parameters_) {
+              insert_tensor(p.metadata_);
+              if (p.grad_metadata_.has_value()) {
+                insert_tensor(*p.grad_metadata_);
               }
             }
+          }
 
-            // torch.optim.Optimizer
-            if (py_call.optimizer_.has_value()) {
-              for (auto& p : py_call.optimizer_->parameters_) {
-                insert_tensor(p.metadata_);
-                if (p.grad_metadata_.has_value()) {
-                  insert_tensor(*p.grad_metadata_);
-                }
-                for (auto& state_i : p.state_) {
-                  insert_tensor(state_i.second);
-                }
+          // torch.optim.Optimizer
+          if (py_call.optimizer_.has_value()) {
+            for (auto& p : py_call.optimizer_->parameters_) {
+              insert_tensor(p.metadata_);
+              if (p.grad_metadata_.has_value()) {
+                insert_tensor(*p.grad_metadata_);
+              }
+              for (auto& state_i : p.state_) {
+                insert_tensor(state_i.second);
               }
             }
-          },
-          [](const auto&) {}));
+          }
+        },
+        [](const auto&) {}));
+  }
+
+  // Step 2) Map TensorImpls to allocations and remove non-Tensor allocations.
+  // --------------------------------------------------------------------------
+  {
+    ska::flat_hash_map<StorageImplData, TensorImplAddress> current;
+    auto update = [&current](RawTensorStoragePair& t, PairType update_type) {
```

Comment:
I do not understand this logic. So,
1) update impladdr for previous same storage tensors only for raw_tensors which are not alloc/free?
2) update raw_tensor (`t`)'s impladdr  only for alloc/free tensor inputs? 
3) Dont only alloc/free tensors have `NoTensorImpl`? three if conditions are very confusing to me.

Reply:
This stems from the fact that Allocation instrumentation just gives us a raw pointer. So we walk forward through the events keeping track of the last `TensorImpl*` we saw for each storage. Then when we get to a deallocation, we can check that list and if `ptr` is in `current` we know it is the address for a storage and we know the corresponding `TensorImpl*`. The same is true for allocations, except we have to walk the events in reverse order. Does that help? (If so, I'll add a comment to that effect.)